### PR TITLE
[TK-08002 TK-08017] req/res handling in tx2_api

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/holochain/pull/720

[Video Walkthrough: Tx2 Api (5:42)](https://www.youtube.com/watch?v=VuAkRUqYj4A)

- Adds request/response handling layer on top of the tx2_frontend traits / structs
- Adds codec parsing/encoding handling layer on top of the tx2_frontend traits / structs

### CODE REVIEW

File: crates/kitsune_p2p/types/src/tx2/tx2_api.rs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        Self(HashMap::new())
+    }
+
+    pub fn insert(&mut self, con: ConHnd, msg_id: u64, s_res: RSend<C>, timeout: KitsuneTimeout) {
+        self.0.insert((con, msg_id), (s_res, timeout));
+    }
+
+    pub fn respond(&mut self, con: ConHnd, msg_id: u64, c: C) {
+        if let Some((s_res, _)) = self.0.remove(&(con, msg_id)) {
+            let _ = s_res.send(Ok(c));
```
Comment (freesig) [inline crates/kitsune_p2p/types/src/tx2/tx2_api.rs:35]: Is the `let _` an error? Does it matter?

File: crates/kitsune_p2p/types/src/tx2/tx2_api.rs [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    ) -> impl std::future::Future<Output = KitsuneResult<()>> + 'static + Send {
+        let Tx2Respond { con, msg_id, .. } = self;
+        async move {
+            let mut buf = PoolBuf::new();
+            data.encode(&mut buf).map_err(KitsuneError::other)?;
+
+            // ignore errors, the remote con may have closed
+            // but that's their loss.
+            let _ = con.write(MsgId::new(msg_id).as_res(), buf, timeout).await;
+
```
Comment (freesig) [inline crates/kitsune_p2p/types/src/tx2/tx2_api.rs:257]: This seems ok but I wonder if it's worth a trace or debug level log?

**Review by freesig — APPROVED** [review]
Looking good. Few minor questions.
Also +1 for notify soon, I think there's some places that holochain can make use of send and forget.

File: crates/kitsune_p2p/types/src/tx2/tx2_api.rs (hunk 1 above)
Comment (neonphog) [reply crates/kitsune_p2p/types/src/tx2/tx2_api.rs:35]: It's normal that a receiver would be dropped when it times out and thus the respond would go nowhere - added a comment.

File: crates/kitsune_p2p/types/src/tx2/tx2_api.rs (hunk 2 above)
Comment (neonphog) [reply crates/kitsune_p2p/types/src/tx2/tx2_api.rs:257]: the fn here already returns a result type, so i just went ahead and bubbled the errors up here.

**Pull request merged**